backtrack/tsp.go

```diff
@@ -0,0 +1,46 @@
+// tsp.go
+// description: solution of the traveling salesman problem
+// details:
+// Given a list of cities and the distances between each pair of cities,
+// what is the shortest possible route that
+// visits each city exactly once and returns to the origin city?
+// wikipedia: https://en.wikipedia.org/wiki/Travelling_salesman_problem
+// author: [sixwaaaay](https://github.com/sixwaaaay)
+// see tsp_test.go
+
+package backtrack
+
+import "math"
+
+// TSP is a solution of the traveling salesman problem
+// we assume that the graph is connected
+func TSP(graph [][]int) (int, []int) {
+	n := len(graph)
+	// x is the current solution, bestX is the best solution found so far
+	x, bestX := make([]int, n), make([]int, n)
+	for i := 0; i < n; i++ {
+		x[i] = i
+	} // initialize x as the permutation of the numbers from 0 to n-1
+	bestCost := math.MaxInt32 // best cost
+	var tsp func(int, int)
+	tsp = func(d int, cost int) { // depth, cost so far
```

Comment:
Same problem with concurrent access applies here... Not sure if this is ideal. Can be solved similarly by passing values by reference?

This pattern exists somewhere I read recently but I personally don't think its a good way since the closure is passing references anyways so having a separate function to receive these values by reference would be best and in fact quite clear as to what is happening (in my opinion)